### PR TITLE
Wrap auto-generated client with gem to add cache support

### PR DESCRIPTION
**Pull request by ethax-ross**
- Rename /gem to /auto-generated-gem

There will be multiple Rubygems in this repository, this makes it clear which is the auto-generated gem create from swagger-codegen.

- Bump version

- Add gem wrapping auto-generated client

The gem monkey patches the auto-generated client to use Faraday (which supports ETag HTTP caching) instead of Typheous.

Client-side caching of all HTTP GET requests is also added (with a TTL of 10 minutes).

Augments the auto generated client with retry logic to ensure failed requests are retried.

- Update workflows to run tests on both gems

- Update README reference new gem and configuration options

### CODE REVIEW

File: api-client/api-client.gemspec [hunk 1]
```diff
@@ -0,0 +1,39 @@
+require_relative 'lib/api/client/version'
+
+Gem::Specification.new do |spec|
+  spec.name          = "api-client"
```
Comment (jebw) [inline api-client/api-client.gemspec:4]: This should be named something which we can publish to rubygems in the future if we wish to (or even to a future private DfE gemserver)

```ruby
spec.name  = "get_into_teaching_api_client_faraday"
```

or

```ruby
spec.name  = "get_into_teaching_api_client_enhanced"
```

**Review by jebw — CHANGES_REQUESTED** [review]
Looks good overall - I think the name should reference the gem it builds off of

Separate to that, I've realised the 2 gem approach makes us depend upon both typhoeus and farady. No strong feelings on that though

An alternative approach would be to just add a monkey patch file and modify the gemspec. Then be careful about what gets committed when rebuilding

When it needs to be regenerated, its then just a case of 

1. rm everything
2. regenerate
3. git checkout `*.gemspec`
4. git checkout lib/monkey_patch_file.rb
5. git commit changes

File: api-client/api-client.gemspec (hunk 1 above)
Comment (ethax-ross) [reply api-client/api-client.gemspec:4]: Yeah good point, I'll get that updated 👍

**Review by jebw — APPROVED** [review]
LGTM

**Pull request merged**